prompting/miners/hf_miner.py

```diff
@@ -76,72 +81,137 @@ def __init__(self, config=None):
+        async def _forward(
+            self,
+            prompt: str,
+            thread: Thread,
+            init_time: float,
+            timeout_threshold: float,
+            streamer: CustomTextIteratorStreamer,
+            send: Send,
+        ):
+            """_summary_
+
+            Args:
+                prompt (str): The received message (challenge) in the synapse. For logging.
+                thread (Thread): A background thread that is reponsible for running the model.
+                init_time (float): Initial time of the forward call. For timeout calculation.
+                timeout_threshold (float): The amount of time that the forward call is allowed to run. If timeout is reached, streaming stops and
+                    validators recieve a partial response.
+                streamer (CustomTextIteratorStreamer): Iterator that holds tokens within a background Queue to be returned when sampled.
+                send (Send): bittensor aiohttp send function to send the response back to the validator.
+            """
             bt.logging.debug(f"📧 Message received, forwarding synapse: {synapse}")
 
-            prompt = synapse.messages[-1]
-            bt.logging.debug(f"💬 Querying {self.model_id}: {prompt}")
-
-            response = HuggingFaceLLM(
+            buffer = []
+            temp_completion = ""  # for wandb logging
+            timeout_reached = False
+
+            try:
+                for token in streamer:
+                    buffer.append(token)
+
+                    if time.time() - init_time > timeout_threshold:
+                        bt.logging.debug(f"⏰ Timeout reached, stopping streaming")
+                        timeout_reached = True
+                        break
+
+                    if len(buffer) == self.config.neuron.streaming_batch_size:
+                        joined_buffer = "".join(buffer)
+                        temp_completion += joined_buffer
+                        bt.logging.debug(f"Streamed tokens: {joined_buffer}")
+
+                        await send(
+                            {
+                                "type": "http.response.body",
+                                "body": joined_buffer.encode("utf-8"),
+                                "more_body": True,
+                            }
+                        )
+                        buffer = []
+
+                if (
+                    buffer and not timeout_reached
+                ):  # Don't send the last buffer of data if timeout.
+                    joined_buffer = "".join(buffer)
+                    temp_completion += joined_buffer
+                    bt.logging.debug(f"Streamed tokens: {joined_buffer}")
+
+                    await send(
+                        {
+                            "type": "http.response.body",
+                            "body": joined_buffer.encode("utf-8"),
+                            "more_body": False,
+                        }
+                    )
```

Comment:
I see this pattern across all the miners, maybe it's too late for thinking in a more generic abstraction but could be something to have in mind in the next iterations